docs/cvl/v5-changes.md

```diff
@@ -193,6 +194,52 @@ is the main contract being verified, then passing `--contract C` will cause
 method variables to be instantiated in the same way the would have in older
 versions.
 
+Disallow Solidity function calls in quantified expressions
+----------------------------------------------------------
+
+Starting with `certora-cli` version 5.0, the Prover no longer supports
+making Solidity contract method calls in quantified expression bodies by
+default. For example, given the simple contract below, you can no longer
+use the `method` `foo()` in a quantified expression body.
+
+```{code-block} solidity
+contract example {
+   function foo(uint256 i) public pure returns (uint256) {
+       return i;
+   } 
+}
+```
+
+```{code-block} cvl
+:emphasize-lines: 4
+
+rule for_all() {
+    // Using foo(i) in the quantified body will now cause the prover to
+    // generate a type-checking error.
+    require (forall uint256 i . i == foo(i));
+    assert false, "Prover will generate an error before this line";
+}
+```
+
+In the example rule `for_all`, the prover will now generate an error similar
+to the following:
+
+```text
+Error in spec file (test2.spec:8:36): Contract function calls such as foo(i)
+are disallowed inside quantified formulas.
+```
+
+Using Solidity `method` calls in the quantified expressions introduced an
+element of instability to quantifier expressions that you can better
+overcome using ghosts, direct storage access, and other access patterns.
```

Comment:
There are two ideas in this paragraph.  The first is why we made the change, and the second is how to work around it.  I would probably either remove the first part or move it to the first paragraph in the section.  I would expand on the second point; the advice here isn't really actionable.  I would probably demonstrate how to rewrite the `foo` example above with both ghosts and direct storage access.

E.g. something like

"In most simple cases, you can replace the Solidity function call with either a {ref}`direct storage access <...>` or a {ref}`ghost <ghosts>`.  For example, the above function `foo` simply returns the storage variable `i`, so the `require` statement in the `for_all` rule could be replaced by `require (forall uint i . i == currentContract.i)`.  Alternatively, you could create a ghost variable that tracks `foo`:

```cvl
...
```

Reply:
I realize on second glance that this doesn't actually describe what `foo` does, and maybe this is harder to do for that example, so I might change the example a bit.

Reply:
I reworked the example and removed the explanation for the change.